File: protocol/src/main/resources/javascript/ProtocolManagerTemplate.js
${protocol_imports}

const protocols = new Map();

const ProtocolManager = {};

// initProtocol
${protocol_manager_registrations}

ProtocolManager.getProtocol = function getProtocol(protocolId) {
    const protocol = protocols.get(protocolId);
    if (protocol === null) {
        throw new Error('[protocolId:' + protocolId + '] not exist');
    }
    return protocol;
};

ProtocolManager.write = function write(buffer, packet) {
    const protocolId = packet.protocolId();
    buffer.writeShort(protocolId);
    const protocol = ProtocolManager.getProtocol(protocolId);
    protocol.write(buffer, packet);
};

ProtocolManager.read = function read(buffer) {
    const protocolId = buffer.readShort();
    const protocol = ProtocolManager.getProtocol(protocolId);
    const packet = protocol.read(buffer);
    return packet;
};

export default ProtocolManager;

File: protocol/src/main/resources/scala/ProtocolManagerTemplate.scala
${protocol_root_path}
${protocol_imports}
import scala.collection.mutable

object ProtocolManager {
  val MAX_PROTOCOL_NUM: Short = Short.MaxValue
  val protocols = new Array[IProtocolRegistration](MAX_PROTOCOL_NUM)
  val protocolIdMap = mutable.Map[Class[_], Short]()

  def initProtocol(): Unit = {
    // initProtocol
    ${protocol_manager_registrations}
  }

  def getProtocolId(clazz: Class[_]): Short = protocolIdMap.getOrElse(clazz, -1)

  def getProtocol(protocolId: Short): IProtocolRegistration = {
    val protocol = protocols(protocolId)
    if (protocol == null) throw new RuntimeException("[protocolId:" + protocolId + "] not exist")
    protocol
  }

  def write(buffer: ByteBuffer, packet: Any): Unit = {
    val protocolId = getProtocolId(packet.getClass)
    // write protocol id to buffer
    buffer.writeShort(protocolId)
    // write packet
    getProtocol(protocolId).write(buffer, packet)
  }

  def read(buffer: ByteBuffer): Any = {
    val protocolId = buffer.readShort
    getProtocol(protocolId).read(buffer)
  }
}